xds/src/main/java/io/grpc/xds/LoadBalancerConfigFactory.java

```diff
@@ -0,0 +1,326 @@
+ *
+ * <p>The new approach is to set the load_balancing_policy field that contains both the policy
+ * selection as well as any supporting configuration data. Providing a list of acceptable policies
+ * is also supported. Note that if this field is used, it will override any configuration set using
+ * the legacy approach. The new configuration approach is explained in detail in the <a href="
+ * https://github.com/grpc/proposal/blob/master/A52-xds-custom-lb-policies.md">Custom LB Policies
+ * gRFC</a>
+ */
+class LoadBalancerConfigFactory {
+
+  private static XdsLogger logger = XdsLogger.withLogId(
+      InternalLogId.allocate("xds-client-lbconfig-factory", null));
+
+  static final String ROUND_ROBIN_FIELD_NAME = "round_robin";
+
+  static final String RING_HASH_FIELD_NAME = "ring_hash_experimental";
+  static final String MIN_RING_SIZE_FIELD_NAME = "minRingSize";
+  static final String MAX_RING_SIZE_FIELD_NAME = "maxRingSize";
+
+  static final String LEAST_REQUEST_FIELD_NAME = "least_request_experimental";
+  static final String CHOICE_COUNT_FIELD_NAME = "choiceCount";
+
+  static final String WRR_LOCALITY_FIELD_NAME = "wrr_locality_experimental";
+  static final String CHILD_POLICY_FIELD = "childPolicy";
+
+  static final String TYPE_URL_PREFIX_REGEX = "type\\.googleapis\\.com/";
+
+  /**
+   * Factory method for creating a new {link LoadBalancerConfigConverter} for a given xDS {@link
+   * Cluster}.
+   *
+   * @throws ResourceInvalidException If the {@link Cluster} has an invalid LB configuration.
+   */
+  static ImmutableMap<String, ?> newConfig(Cluster cluster, boolean enableLeastRequest)
+      throws ResourceInvalidException {
+    // The new load_balancing_policy will always be used if it is set, but for backward
+    // compatibility we will fall back to using the old lb_policy field if the new field is not set.
+    if (cluster.hasLoadBalancingPolicy()) {
+      try {
+        return LoadBalancingPolicyConverter.convertToServiceConfig(cluster.getLoadBalancingPolicy(),
+            0);
+      } catch (MaxRecursionReachedException e) {
+        throw new ResourceInvalidException("Maximum LB config recursion depth reached");
+      }
+    } else {
+      return LegacyLoadBalancingPolicyConverter.convertToServiceConfig(cluster, enableLeastRequest);
+    }
+  }
+
+  /**
+   * Builds a service config JSON object for the ring_hash load balancer config based on the given
+   * config values.
+   */
+  private static ImmutableMap<String, ?> buildRingHashConfig(Long minRingSize, Long maxRingSize) {
+    ImmutableMap.Builder<String, Object> configBuilder = ImmutableMap.builder();
+    if (minRingSize != null) {
+      configBuilder.put(MIN_RING_SIZE_FIELD_NAME, minRingSize.doubleValue());
+    }
+    if (maxRingSize != null) {
+      configBuilder.put(MAX_RING_SIZE_FIELD_NAME, maxRingSize.doubleValue());
+    }
+    return ImmutableMap.of(RING_HASH_FIELD_NAME, configBuilder.build());
```

Comment:
There's been a recent change to master to swap to `ImmutableMap.Builder.buildOrThrow()`, which is functionally equivalent to `build()`. We should probably use the new method in this PR.